### PR TITLE
Agent lb on svm

### PR DESCRIPTION
**Pull request by DaanHoogland**
## Description

If a host is not in the list of hosts and an agent load balancing list is required for it, add it to the list to get a valid MSlist.

Fixes: #3722

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java [hunk 1]
```diff
@@ -136,17 +144,48 @@ public int compare(Long x, Long y) {
...
         for (final Host host : allHosts) {
-            if (host == null || host.getResourceState() == ResourceState.Creating || host.getResourceState() == ResourceState.Error) {
-                continue;
+            conditionallyAddHost(agentBasedHosts, host);
+        }
+        return agentBasedHosts;
+    }
+
+    private void conditionallyAddHost(List<Host> agentBasedHosts, Host host) {
+        if (host == null) {
```
Comment (rhtyd) [inline server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: We can do null checks and if host is removed or is some unsupported state like error etc.?

File: server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java [hunk 2]
```diff
@@ -136,17 +144,48 @@ public int compare(Long x, Long y) {
...
+            return;
+        }
+        if (host.getResourceState() == ResourceState.Creating) {
+            if (LOG.isTraceEnabled()) {
+                LOG.trace(String.format("host is in creating state, not adding to the host list, (id = %s)", host.getUuid()));
             }
+            return;
         }
-        return agentBasedHosts;
+        if (host.getResourceState() == ResourceState.Error) {
```
Comment (rhtyd) [inline server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: oh I see it's already here, we could probably do something like `item in array` where array is list of unsupported states?

Comment (DaanHoogland) [reply server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: refactorred have a look, @rhtyd . I think it is cosmetic but it does make a point.

File: server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java [hunk 3]
```diff
@@ -136,17 +146,54 @@ public int compare(Long x, Long y) {
...
             }
-            if (host.getType() == Host.Type.Routing || host.getType() == Host.Type.ConsoleProxy || host.getType() == Host.Type.SecondaryStorage || host.getType() == Host.Type.SecondaryStorageVM) {
-                if (host.getHypervisorType() != null && host.getHypervisorType() != Hypervisor.HypervisorType.KVM && host.getHypervisorType() != Hypervisor.HypervisorType.LXC) {
-                    continue;
-                }
-                agentBasedHosts.add(host);
+            return;
+        }
+
+        ResourceState[] allowedStates = new ResourceState[]{
```
Comment (rhtyd) [inline server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: Why not use Arrays.asList()?

File: server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java [hunk 4]
```diff
@@ -136,17 +146,54 @@ public int compare(Long x, Long y) {
...
+                ResourceState.Enabled,
+                ResourceState.Maintenance,
+                ResourceState.Disabled,
+                ResourceState.ErrorInMaintenance,
+                ResourceState.PrepareForMaintenance
+        };
+        // so the remaining ResourceState[] disallowedStates = new ResourceState[]{ResourceState.Creating, ResourceState.Error};
+        if (binarySearch(allowedStates, host.getResourceState()) < 0) {
```
Comment (rhtyd) [inline server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: Why not use `list.contains()`?

Comment (DaanHoogland) [reply server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: seems like less cpu power/more code?

Comment (DaanHoogland) [reply server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: @rhtyd ?

Comment (rhtyd) [reply server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: @DaanHoogland premature optimisation is... nope there won't be a lot of performance gain. If you have to you can use `Collections.binarySearch()`

**Comment by rhtyd** [comment]
@DaanHoogland replied, let me know if I've anything else.

File: server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java (hunk 4 above)
Comment (DaanHoogland) [reply server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java]: ok, it felt quite natural to me, but will use an EnumSet.contains().

**Comment by DaanHoogland** [comment]
ping @andrijapanicsb this can do with a second pair of eyes and some tests

**Comment by andrijapanicsb** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
Tested manually by reproducing the issue on clean 4.13, confirming CPVM won't start with the exceptions below...:

> 2020-01-14 14:02:44,379 DEBUG [c.c.a.ConsoleProxyAlertAdapter] (consoleproxy-1:ctx-a32cef4f) (logid:5d7a9f58) received console proxy alert
2020-01-14 14:02:44,380 DEBUG [c.c.a.ConsoleProxyAlertAdapter] (consoleproxy-1:ctx-a32cef4f) (logid:5d7a9f58) Console proxy creation failure, zone: ref-trl-451-x-M7-apanic
2020-01-14 14:02:44,381 WARN  [c.c.a.AlertManagerImpl] (consoleproxy-1:ctx-a32cef4f) (logid:5d7a9f58) AlertType:: 10 | dataCenterId:: 1 | podId:: null | clusterId:: null | message:: Console proxy creation failure. zone: ref-trl-451-x-M7-apanic, error details: fromIndex = -1
2020-01-14 14:02:44,387 WARN  [c.c.v.SystemVmLoadScanner] (consoleproxy-1:ctx-a32cef4f) (logid:5d7a9f58) Unexpected exception fromIndex = -1
java.lang.IndexOutOfBoundsException: fromIndex = -1
        at java.util.SubList.<init>(AbstractList.java:620)
        at java.util.RandomAccessSubList.<init>(AbstractList.java:775)

Upgrade to this one, and CPVM starts fine.

**Comment by andrijapanicsb** [comment]
LGTM

**Pull request merged**